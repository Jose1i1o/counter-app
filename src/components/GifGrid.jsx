import { useState, useEffect } from "react";
import GifGridItem from "./GifGridItem";

const GifGrid = ({ category }) => {
  
  // const [counter, setCounter] = useState(0);

  const [ images, setImages ] = useState([]);

  useEffect(() => {
    getGifs();
  },[]) // This tells to render this component only once. Equals componentDidMount

  const getGifs = async() => {

    const url = 'https://api.giphy.com/v1/gifs/search?q=Rick and Morty&api_key=EhdRmrgVkNVGNw9aYoKS90hAJgBbPYz5&limit=20';
    const response = await fetch(url);
    const { data } = await response.json();

    const gifs = data.map( img => {
      return {
        id: img.id,
        title: img.title,
        url: img.images?.downsized_medium.url
      }
    })
    // console.log(data);
    // console.log(gifs);
    setImages( gifs )
  }
  // getGifs();

  return (
    <>
      <h3>{ category }</h3>
      {/* <ol> */}
        {/* {
         images.map(({ id, title }) => (
           <li key={ id }>{ title }</li>
         ))
        } */}
      {/* </ol> */}
        {
         images.map(img  => (
           <GifGridItem 
              key={ img.id }
              { ...img }
           />
         ))
        }
      
      {/* Use this to show how this components renders each time the button is clicked until useEffect is used
      <h3>{ counter }</h3>
      <button onClick={() => setCounter(counter + 1)}>Add +1</button> */}
    </>
  )
}

export default GifGrid;
