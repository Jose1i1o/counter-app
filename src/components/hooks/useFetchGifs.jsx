// name it with use to indicate it is a hook
import { useState, useEffect } from "react";
import getGifs from "../../helpers/getGifs";

const useFetchGifs = ( category ) => {

    const [state, setState] = useState({
      data: [],
      loading: true
    });

    // effects can not be async because they are expecting something synchronous
    useEffect(() => {
      getGifs( category )
        .then( images => {
          // setTimeout(() => {
            setState({
            data: images,
            loading: false
          })
        // }, 500)
      })
    }, [ category ])

  return state;

}


export default useFetchGifs;