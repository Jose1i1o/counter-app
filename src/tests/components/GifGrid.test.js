// Enzyme not available for React 18. Tested with React 17

import React from 'react';
import '@testing-library/jest-dom';

import { shallow } from 'enzyme';
import GifGrid from '../../components/GifGrid';
import useFetchGifs from '../../hooks/useFetchGifs';

// I need both this .mock() and fake the data as in line 33
jest.mock('../../hooks/useFetchGifs');

describe('Testing <GifGrid />', () => {

    const category = 'X-men';


    test('to display correctly', () => {

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow( <GifGrid category={ category } /> );
        expect( wrapper ).toMatchSnapshot();

    });

    test('show items when images are loaded with useFetchGifs', () => {

        // We can simulate data is received
        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/any/item.jpg',
            title: 'Anything'
        },
        {
            id: '123',
            url: 'https://localhost/any/item.jpg',
            title: 'Anything'
        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow( <GifGrid category={ category } /> );

        // expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe(false);
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length );
    })
})

