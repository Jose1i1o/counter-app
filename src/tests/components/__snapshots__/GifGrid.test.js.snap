// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <GifGrid /> to display correctly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
     
    X-men
     
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
